sympy.polys.polyerrors.HeuristicGCDFailed: no luck

```python
from sympy import *

x, y = symbols('x, y')

expr = ((2*y**3*sin(x/y)**2 + x)**2*(y*(-6*y**2*sin(x/y)**2 +
    4*y*x*sin(x/y)*cos(x/y))/(2*y**3*sin(x/y)**2 + x)**2 +
    1/(2*y**3*sin(x/y)**2 + x))/(4*y*(2*y**2*(3*y*sin(x/y) -
        2*x*cos(x/y))**2*sin(x/y)**2/(2*y**3*sin(x/y)**2 + x) - 3*y*sin(x/y)**2
        + 4*x*sin(x/y)*cos(x/y) - (3*y*sin(x/y) - 2*x*cos(x/y))*sin(x/y) +
        x**2*sin(x/y)**2/y - x**2*cos(x/y)**2/y)))

cancel(expr)
```
This gives:
```python
In [2]: cancel(expr)
---------------------------------------------------------------------------
HeuristicGCDFailed                        Traceback (most recent call last)
<ipython-input-2-f5a0e740cc3b> in <module>
----> 1 cancel(expr)

~/current/sympy/sympy/sympy/polys/polytools.py in cancel(f, *gens, **args)
   6729             return f.xreplace(dict(reps))
   6730 
-> 6731     c, (P, Q) = 1, F.cancel(G)
   6732     if opt.get('polys', False) and not 'gens' in opt:
   6733         opt['gens'] = R.symbols

~/current/sympy/sympy/sympy/polys/rings.py in cancel(self, g)
   2222 
   2223         if not (domain.is_Field and domain.has_assoc_Ring):
-> 2224             _, p, q = f.cofactors(g)
   2225         else:
   2226             new_ring = ring.clone(domain=domain.get_ring())

~/current/sympy/sympy/sympy/polys/rings.py in cofactors(f, g)
   2138 
   2139         J, (f, g) = f.deflate(g)
-> 2140         h, cff, cfg = f._gcd(g)
   2141 
   2142         return (h.inflate(J), cff.inflate(J), cfg.inflate(J))

~/current/sympy/sympy/sympy/polys/rings.py in _gcd(f, g)
   2171             return f._gcd_QQ(g)
   2172         elif ring.domain.is_ZZ:
-> 2173             return f._gcd_ZZ(g)
   2174         else: # TODO: don't use dense representation (port PRS algorithms)
   2175             return ring.dmp_inner_gcd(f, g)

~/current/sympy/sympy/sympy/polys/rings.py in _gcd_ZZ(f, g)
   2176 
   2177     def _gcd_ZZ(f, g):
-> 2178         return heugcd(f, g)
   2179 
   2180     def _gcd_QQ(self, g):

~/current/sympy/sympy/sympy/polys/heuristicgcd.py in heugcd(f, g)
     78                 h, cff, cfg = domain.cofactors(ff, gg)
     79             else:
---> 80                 h, cff, cfg = heugcd(ff, gg)
     81 
     82             h = _gcd_interpolate(h, x, ring)

~/current/sympy/sympy/sympy/polys/heuristicgcd.py in heugcd(f, g)
     78                 h, cff, cfg = domain.cofactors(ff, gg)
     79             else:
---> 80                 h, cff, cfg = heugcd(ff, gg)
     81 
     82             h = _gcd_interpolate(h, x, ring)

~/current/sympy/sympy/sympy/polys/heuristicgcd.py in heugcd(f, g)
     78                 h, cff, cfg = domain.cofactors(ff, gg)
     79             else:
---> 80                 h, cff, cfg = heugcd(ff, gg)
     81 
     82             h = _gcd_interpolate(h, x, ring)

~/current/sympy/sympy/sympy/polys/heuristicgcd.py in heugcd(f, g)
    116         x = 73794*x * domain.sqrt(domain.sqrt(x)) // 27011
    117 
--> 118     raise HeuristicGCDFailed('no luck')
    119 
    120 def _gcd_interpolate(h, x, ring):

HeuristicGCDFailed: no luck
```

There are lots of test failures if gmpy is installed:
```console
$ pytest sympy/core/ sympy/polys -n4 -d -m 'not slow'
...
                                                                       DO *NOT* COMMIT!                                                                       
================================================================== short test summary info ===================================================================
FAILED sympy/core/tests/test_arit.py::test_issue_8247_8354 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/core/tests/test_arit.py::test_Add_is_negative_positive - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/core/tests/test_expr.py::test_equals - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_euclidtools.py::test_dup_gcd - assert 3654318780237...73459383222272 == 3654318780237...30321821976049
FAILED sympy/polys/tests/test_factortools.py::test_dup_qq_i_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dmp_qq_i_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dup_zz_i_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dmp_zz_i_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dup_ext_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dmp_ext_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dup_factor_list - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_factortools.py::test_dmp_factor_list - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_issue_19760 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_partfrac.py::test_apart_extension - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_partfrac.py::test_apart_list - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_partfrac.py::test_assemble_partfrac_list - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polyroots.py::test_issue_14522 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_issue_20427 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_minimal_polynomial - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_factor_large - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_minimal_polynomial_issue_19732 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_minimal_polynomial_hi_prec - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_sturm - sympy.polys.polyerrors.HeuristicGCDFailed: no luck
FAILED sympy/polys/tests/test_polyroots.py::test_issue_21287 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_minpoly_compose - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_cancel - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_factor - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polyroots.py::test_roots_cyclotomic - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_primitive_element - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polyroots.py::test_roots0 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_rings.py::test_PolyElement_sturm - sympy.polys.polyerrors.HeuristicGCDFailed: no luck
FAILED sympy/polys/tests/test_numberfields.py::test_field_isomorphism - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polyroots.py::test_roots_mixed - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_minpoly_fraction_field - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_polytools.py::test_Poly__unify - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_numberfields.py::test_issue_14831 - TypeError: PythonMPQ() requires numeric or string argument
FAILED sympy/polys/tests/test_solvers.py::test_solve_lin_sys_6x6_2 - sympy.polys.polyerrors.HeuristicGCDFailed: no luck
======================================= 37 failed, 2922 passed, 69 skipped, 43 xfailed, 1 xpassed in 165.32s (0:02:45) =======================================
```

--- a/sympy/external/gmpy.py
+++ b/sympy/external/gmpy.py
@@ -97,5 +97,5 @@ else:
     MPZ = int
     MPQ = PythonMPQ
 
-    factorial = mlib.ifac
-    sqrt = mlib.isqrt
+    factorial = lambda x: int(mlib.ifac(x))
+    sqrt = lambda x: int(mlib.isqrt(x))

